File: app/src/test/java/io/oldering/tvfoot/red/util/Fixture.java
package io.oldering.tvfoot.red.util;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.google.gson.stream.JsonReader;
import io.oldering.tvfoot.red.data.entity.Match;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.lang.reflect.Type;
import java.util.List;
import javax.inject.Inject;
import timber.log.Timber;

public class Fixture {
  private Gson gson;

  @Inject public Fixture(Gson gson) {
    this.gson = gson;
  }

  private Type listMatchType = new TypeToken<List<Match>>() {
  }.getType();

  public List<Match> anyMatches() {
    return anyObject("matches_sample.json", listMatchType);
  }

  public List<Match> anyNextMatches() {
    return anyObject("matches_next_sample.json", listMatchType);
  }

  private <T> T anyObject(String filename, Type typeOfT) {
    InputStream inputStream = this.getClass().getClassLoader().getResourceAsStream(filename);

    try {
      JsonReader reader = new JsonReader(new InputStreamReader(inputStream, "UTF-8"));

      T object = gson.fromJson(reader, typeOfT);

      reader.close();

      return object;
    } catch (IOException e) {
      Timber.e(e);
      throw new RuntimeException("fixture " + filename + " failed");
    }
  }
}

File: app/src/test/java/com/benoitquenaudon/tvfoot/red/testutil/Fixture.kt
package com.benoitquenaudon.tvfoot.red.testutil

import com.benoitquenaudon.tvfoot.red.app.data.entity.Match
import com.squareup.moshi.JsonReader
import com.squareup.moshi.Moshi
import com.squareup.moshi.Types
import okio.Okio
import timber.log.Timber
import java.io.IOException
import java.lang.reflect.Type
import javax.inject.Inject

class Fixture @Inject constructor(private val moshi: Moshi) {

  private val listMatchType =
      Types.newParameterizedType(List::class.java, Match::class.java)

  fun anyMatches(): List<Match> {
    return anyObject("matches_sample.json", listMatchType)
  }

  fun anyMatchesShort(): List<Match> {
    return anyObject("matches_sample_short.json", listMatchType)
  }

  fun anyMatch(): Match {
    return anyObject("match_sample.json", Match::class.java)
  }

  fun anyNextMatches(): List<Match> {
    return anyObject("matches_next_sample.json", listMatchType)
  }

  private fun <T> anyObject(filename: String, typeOfT: Type): T {
    val inputStream = this.javaClass.classLoader.getResourceAsStream(filename)

    try {
      val reader = JsonReader.of(Okio.buffer(Okio.source(inputStream)))
      val adapter = moshi.adapter<T>(typeOfT)
      val result = adapter.fromJson(reader)

      reader.close()

      return result ?: throw IllegalStateException("no data for $filename and $typeOfT")
    } catch (e: IOException) {
      Timber.e(e)
      throw RuntimeException("fixture $filename failed")
    }
  }
}
